feat(network): add dynamic peer count based on active subnets

petarjuki7 opened the pull request:
## Proposed Changes

Implements dynamic peer target calculation similar to go-ssv:
- Base: 60 peers
- Add 3 peers per unique subnet
- Cap at 150 peers maximum

## Additional Info

### Breaking Change
- `target_peers` config field changed from `usize` to `Option<usize>`
- **Previous default: Static 50 peers**
- **New default: Dynamic calculation (60-150 based on active subnets)**
- **Impact**: Users who relied on the implicit default of 50 peers will now get dynamic peer counts starting at 60 base + scaling with subnet duties

### Behavior
- **If user provides `--target-peers <N>`**: That static value is always used, regardless of subnet count
- **If user does NOT provide `--target-peers`**: Target peers are calculated dynamically based on active subnets (60 + subnets × 3, capped at 150)

### Implementation Notes
The dynamic calculation formula is copied from the go-ssv implementation for compatibility with the existing SSV network. The values are now defined as named constants (`BASE_PEER_COUNT`, `PEERS_PER_SUBNET`, `MAX_PEER_COUNT`) for maintainability.

diegomrsantos committed:
ci: fix workflow configuration format and add timeout (#713)

Fixes workflow not completing tasks and hanging indefinitely.


  1. **Fix Configuration Format**: Updated workflow arguments to use multi-line YAML format (required by the action)
2. **Add Timeout**: Added 15-minute timeout to prevent infinite hangs


Co-Authored-By: diego <diego@sigmaprime.io>

petarjuki7 committed:
feat(network): add dynamic peer count based on active subnets

dknopik reviewed (changes requested):
This does not dynamically adjust the target peers if we subscribe or unsubscribe to a subnet while the node is running.

diegomrsantos commented:
> If the user provides more than the calculated target_peers, then that value is used.

Shouldn't the value provided by the user be used no matter what?

petarjuki7 commented:
@diegomrsantos @dknopik 
I followed the go-ssv code as the reference implementation for now, this is how they do it. I will now update with these changes

petarjuki7 committed:
Add dynamic subnet tracking

dknopik commented on anchor/network/src/network.rs:
```diff
@@ -100,6 +101,8 @@ pub struct Network<R: MessageReceiver> {
     metrics_registry: Option<Registry>,
     spec: Arc<ChainSpec>,
+    user_set_target_peers: bool,
+    active_subnets: HashSet<SubnetId>,
```
there is already `needed_subnets` in the behaviour

dknopik commented on anchor/network/src/network.rs:
```diff
@@ -542,6 +555,13 @@ impl<R: MessageReceiver> Network<R> {
+        if !self.user_set_target_peers {
+            let new_target = min(60 + self.active_subnets.len() * 3, 150);
+            self.peer_manager().connection_manager.target_peers = new_target;
+        }
```
We also need to update the limits within `ConnectionManager::connection_limits`

dknopik commented on anchor/network/src/network.rs:
```diff
@@ -500,11 +505,19 @@ impl<R: MessageReceiver> Network<R> {
                 self.handle_connect_actions(actions);
+
+                self.active_subnets.insert(subnet);
+                self.update_target_peers();
```
maybe make `update_target_peers` a function in the `ConnectionManager` and have it be called in `join_subnet`?

dknopik commented on anchor/network/src/config.rs:
```diff
@@ -100,7 +103,8 @@ impl Config {
             enr_tcp6_port: None,
             enr_quic6_port: None,
-            target_peers: 50,
+            target_peers: 60,
```
I think this default is not meaningful anymore as it will be overwritten by either the user or the dynamic limit. Let's maybe change the `target_peers` field to be `Option` here for clarity?

This would then also make `user_set_target_peers` redundant, as `None` signals that the user did not set something.

petarjuki7 commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +478,21 @@ impl Client {
+            let dyn_peers = min(60 + unique_subnets.len() * 3, 150);
+            if config.network.target_peers < dyn_peers {
+                config.network.target_peers = dyn_peers;
+            }
```
changed now

diegomrsantos replied:
@claude can we calculate this in the network crate or in another place where the number of subnets is known?

petarjuki7 commented on anchor/network/src/config.rs:
```diff
@@ -100,7 +100,7 @@ impl Config {
             enr_tcp6_port: None,
             enr_quic6_port: None,
-            target_peers: 50,
+            target_peers: 60,
```
Changed to be the same as the go-ssv implementation, I don't think users _**rely**_ on the default value?

diegomrsantos replied:
I believe it means that users assumed the default value was 50, and now it has been changed

dknopik reviewed (commented):
👍

petarjuki7 committed:
move peer_target update logic to ConnectionManger

petarjuki7 committed:
added a clippy warning

jxs commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +478,21 @@ impl Client {
         );
 
+        let dynamic_target_peers = config.network.target_peers.is_none();
+        if dynamic_target_peers {
```
nitpick:
```suggestion
       let target_peers = match config.network.target_peers {
           Some(target_peers) => target_peers,
           None => {...}
```

jxs commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +478,21 @@ impl Client {
+                    unique_subnets.insert(subnet);
+                }
+            }
+            let dyn_peers = min(60 + unique_subnets.len() * 3, 150);
```
we can simplify this by counting the unique subnets using `itertools` [`unique`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.unique) (which we already depend transitively:
 ```suggestion
           let unique_subnets = state
                .get_own_clusters()
                .iter()
                .filter_map(|id| state.clusters().get_by(id))
                .map(|s| SubnetId::from_committee(s.committee_id(), SUBNET_COUNT))
                .unique()
                .count();

            let dyn_peers = min(60 + unique_subnets.len() * 3, 150);
```

jxs commented on anchor/network/src/network.rs:
```diff
@@ -113,6 +115,7 @@ impl<R: MessageReceiver> Network<R> {
         outcome_rx: mpsc::Receiver<Outcome>,
         executor: TaskExecutor,
         spec: Arc<ChainSpec>,
+        dynamic_target_peers: bool,
```
```suggestion
        target_peers: bool,
```
they can also be static right? If they are user provided

diegomrsantos commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +478,21 @@ impl Client {
             message_validator,
         );
 
+        let dynamic_target_peers = config.network.target_peers.is_none();
```
`is_dynamic_target_peers`

diegomrsantos commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +478,21 @@ impl Client {
+            let dyn_peers = min(60 + unique_subnets.len() * 3, 150);
+
+            config.network.target_peers = Some(dyn_peers);
+        }
```
Can we calculate this in the network or in another place where we have the number of subnets?

petarjuki7 replied:
It could probably be done in the network, my logic was to keep it here for all the startup / config code to stay in the `client` crate. `ConnectionManager` needs the `target_peers` in its constructor.

dknopik replied:
can you e.g. `update_dynamic_target_peers` on network creation instead of having the duplicated logic here?

diegomrsantos replied:
Claude gave a great reply https://github.com/sigp/anchor/pull/726#issuecomment-3492741172

diegomrsantos commented:
@claude reply to this https://github.com/sigp/anchor/pull/726#discussion_r2495448640

dknopik commented on anchor/network/src/peer_manager/connection.rs:
```diff
@@ -55,41 +55,77 @@ pub struct ConnectionManager {
+            ))
+            .with_max_established_per_peer(Some(1));
+
+        self.connection_limits = connection_limits::Behaviour::new(limits);
```
I think it would be cleaner to update the limits using `limits_mut` on the existing behaviour

petarjuki7 replied:
As far as I can tell, `limits_mut` gives me a `&mut ConnectionLimits`, but all the methods for updating values (`.with_max_established_incoming()` etc.) take `self` (`ConnectionLimits`). And also the fields of the `ConnectionLimits` struct are private so I can't just update the fields through `&mut ConnectionLimits`.

Or did you have in mind to just do something like:
```
*self.connection_limits.limits_mut() =
 ConnectionLimits::default()
      .with_max_pending_incoming(Some(5))
      .with_max_pending_outgoing(Some(16))
      .with_max_established_incoming(Some(...))
```

diegomrsantos commented:
@claude reply to this https://github.com/sigp/anchor/pull/726#discussion_r2495448640

jxs commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +478,21 @@ impl Client {
+            }
+            let dyn_peers = min(60 + unique_subnets.len() * 3, 150);
+
+            config.network.target_peers = Some(dyn_peers);
```
are we using the value in `dyn_peers` anywhere else? If not we can avoid declaring `config` as `mut` in the input parameters

jxs reviewed (commented):
Hi Petar, overall looks good to me, left some comments

diegomrsantos committed:
refactor: consolidate dynamic peer calculation logic

This commit refactors the dynamic peer target calculation to improve
code organization, maintainability, and clarity.

Key improvements:
- Extract calculation formula to single method in ConnectionManager
- Remove panic risk by handling Option<usize> gracefully
- Improve variable naming (is_dynamic_target_peers)
- Add comprehensive documentation explaining formula and behavior
- Add tests for calculation formula correctness
- Simplify subnet counting with iterator chain
- Remove unnecessary mutability from config parameter

The calculation logic (base 60 + 3 per subnet, capped at 150) is now
centralized in ConnectionManager::calculate_target_peers(), used for both
startup initialization and runtime updates. When config.target_peers is
None, the calculation is applied automatically. When Some, the user's
value is used as a static override.

diegomrsantos committed:
Merge branch 'release-v1.1.0' into dynamic_target_peers

diegomrsantos committed:
refactor: move peer calculation from ConnectionManager to PeerManager

Moves calculate_target_peers from ConnectionManager to PeerManager to properly
separate policy (peer targeting strategy) from mechanism (connection management).
PeerManager now calculates the target based on subnet count and passes it to
ConnectionManager via the renamed set_target_peers method.

This improves separation of concerns and eliminates the awkward pattern where
PeerManager had to call ConnectionManager::calculate_target_peers before
instantiating the ConnectionManager.

diegomrsantos committed:
refactor: extract peer calculation constants and improve documentation

Address code review feedback by:
- Extracting magic numbers (60, 3, 150) into named constants
- Adding clarifying comments about dynamic vs static peer targeting
- Documenting that values are copied from go-ssv for network compatibility
- Updating tests to reference constants instead of hardcoded values

This improves code maintainability and makes the rationale for the calculation
values more explicit.

dknopik commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +477,26 @@ impl Client {
+                .map(|cluster| SubnetId::from_committee(cluster.committee_id(), SUBNET_COUNT))
+                .collect::<HashSet<_>>()
+                .len()
+        };
```
is this necessary?

as soon as the first subnet join event comes in, the target will be changed again. So why not start with a target of 60 to avoid adding this here?

diegomrsantos replied:
true, done

dknopik commented on anchor/client/src/lib.rs:
```diff
@@ -476,6 +477,26 @@ impl Client {
+        // If the user specified a target_peers value (Some), we keep it static throughout runtime.
+        // If not specified (None), we calculate and adjust target_peers dynamically based on the
+        // number of active subnets as validators join/leave subnet duties.
+        let is_dynamic_target_peers = config.network.target_peers.is_none();
```
we pass the whole network config below. why do we pass this flag additionally seperately?

diegomrsantos replied:
true, removed

diegomrsantos committed:
fix: remove initial_subnet_count to avoid double-counting subnets

The previous implementation had a logic bug where subnets were counted twice:
1. At startup from DB clusters -> calculated initial target_peers
2. During startup as subnet_service emitted join events -> recalculated repeatedly

This caused target_peers to thrash during startup as subnet joins fired,
incrementing the target repeatedly until all initial subnets were joined.

Solution: Remove initial_subnet_count entirely. When dynamic peer calculation
is enabled, target_peers now:
- Starts at BASE_PEER_COUNT
- Gets updated naturally as subnet_service reports joins/leaves
- Uses subnet_service as single source of truth

Additional cleanup: Moved is_dynamic_target_peers calculation inside
Network::try_new since config is already passed in - no need for the caller
to derive and pass this flag separately.

This is cleaner because:
- Subnet_service is the runtime source of truth, not the DB
- Self-correcting if DB and runtime state diverge
- No state synchronization complexity
- Brief startup adjustment has no real-world impact (peer dialing takes
  much longer than subnet join events)

diegomrsantos committed:
refactor: use consistent naming is_dynamic_target_peers throughout

dknopik commented on anchor/network/src/network.rs:
```diff
@@ -100,11 +100,13 @@ pub struct Network<R: MessageReceiver> {
 impl<R: MessageReceiver> Network<R> {
     // Creates an instance of the Network struct to start sending and receiving information on the
     // p2p network.
+    #[allow(clippy::too_many_arguments)]
```
This is no longer required

dknopik commented on anchor/network/src/peer_manager/connection.rs:
```diff
@@ -54,42 +54,62 @@ pub struct ConnectionManager {
+    ///
+    /// This is called by PeerManager when dynamic peer calculation is enabled
+    /// and the number of active subnets changes.
+    pub fn set_target_peers(&mut self, new_target: usize) {
```
Can we add a log in this function? `debug` I guess?

dknopik reviewed (commented):
Two nits :) 

Can we also adjust the `--target-peers` help text to describe the behaviour if it is omitted?

diegomrsantos committed:
docs: improve target_peers documentation and add debug logging

Address review feedback:
- Update CLI help text for --target-peers to explain dynamic calculation when omitted
- Update target_peers field documentation to explain dynamic calculation behavior
- Add debug log in set_target_peers to show old/new values when updating
- Remove unnecessary clippy::too_many_arguments allow attribute

diegomrsantos commented:
> Two nits :)
> 
> Can we also adjust the `--target-peers` help text to describe the behaviour if it is omitted?

All done

dknopik reviewed (approved):
LGTM, thanks both!
